### PR TITLE
Adding charm layers

### PR DESCRIPTION
**Pull request by glower**

### CODE REVIEW

**Comment by glower** [comment]
In order to deploy/run SV in our store environment, we need to build charms out of the charm folder. As a reference I took charm folder from the `snapstorage`:
* Main change is that we don't push the payload to the charm repo but get the payload from swift during the deployment
* Instead running the service on the same unit with two different ports we would run it on two different units on the same port
* We need to move the database to our environment first
* Reactive python script is a copy from snapstorage with a changed name
* Makefile is a copy from snapstorage, I removed everything payload related, we will get the payload from the swift

I did test this setup on my local machine with local running juju. Building and running on `focal` works

File: Makefile [hunk 1]
```diff
@@ -130,11 +130,12 @@ build-tarball: install
 	# calculate md5sum and create file with its value
 	md5sum ${TGZ_NAME} > ${TGZ_NAME}.md5
 
+SWIFT_TARGET_NAME=serial-vault-builds/$(VERSION)/serial-vault.tar.gz
```
Comment (glower) [inline Makefile:133]: In order to pull the tarball from swift with `ols.layer` the path needs to have a specific format, so I change it here

File: Makefile [hunk 2]
```diff
@@ -166,9 +167,19 @@ run-docker:
 stop-docker:
 	cd docker-compose && docker-compose kill && docker-compose rm
 
+# essential charm makefile config
+CHARM_NAME = ${SERVICE_NAME}
+CHARM_PUBLISH_REPO ?= lp:~ubuntuone-pqm-team/serial-vault/+git/charm-build
```
Comment (glower) [inline Makefile:172]: This needs to be created

File: charm/Makefile [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	cd $(CHARM_REPO_DIR) && git -c user.name="$(GIT_USERNAME)" -c user.email="$(GIT_EMAIL)"  commit -am "Build of $(CHARM_NAME) from $(TRUNK_REVNO)"
+	cd $(CHARM_REPO_DIR) && git tag "$(shell git rev-parse --short HEAD)"
+	cd $(CHARM_REPO_DIR) && git push origin master --tags
+
+charm-private-publish:
+	git ls-remote -q $(PRIVATE_PUBLISH_REPO) || ( cd $$(mktemp -d); git init; git commit --allow-empty -m 'init'; git push $(PRIVATE_PUBLISH_REPO) master)
+	$(MAKE) charm-publish CHARM_PUBLISH_REPO=$(PRIVATE_PUBLISH_REPO) CHARM_REPO_DIR=$(PRIVATE_CHARM_REPO_DIR)
+
+# get swift credentials
+-include ~/.config/swift/serial-vault
```
Comment (glower) [inline charm/Makefile:128]: We need to have this file with swift credentials in the `serial-vault` environment

File: charm/bundle.yaml.in [hunk 4]
```diff
@@ -0,0 +1,45 @@
+series: focal
+description: Serial-Vault development bundle
+applications:
+    serial-vault-admin:
+        num_units: 1
+        charm: ../dist/focal/serial-vault
+        options:
+            # ols layer config
+            build_label: "%BUILD_LABEL%"
```
Comment (glower) [inline charm/bundle.yaml.in:9]: this is needed to pull the payload from swift automatically

File: charm/serial-vault/layer.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+repo: https://github.com/CanonicalLtd/serial-vault
+includes:
+    - layer:ols-http
+    - layer:ols-pg
+options:
+    ols:
+        service_name: serial-vault
+        config_filename: settings.yaml
+        user: ols
+        tarball_payload: True
```
Comment (glower) [inline charm/serial-vault/layer.yaml:10]: this is needed to pull the payload from swift automatically

File: charm/Makefile [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+export ~/.config/swift/serial-vault
+
+bundle.yaml: ./charm/bundle.yaml.in
+	sed -e 's/%BUILD_LABEL%/${VERSION}/g' \
+		-e 's/%SWIFT_USER_DOMAIN_NAME%/${OS_USER_DOMAIN_NAME}/g' \
```
Comment (Roadmaster) [inline charm/Makefile:133]: This sed processing is horrible, I wonder if we could use actual variable names in bundle.yaml.in and just run it through envsubst (from the `gettext-base` .deb package);

 `envsubst < ./charm/bundle.yaml.in > ./charm/bundle.yaml`

File: charm/packages.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+charm-tools
+python-psycopg2
+libpq-dev
+virtualenv
+language-pack-en-base
+bzr
+python3-swiftclient
```
Comment (Roadmaster) [inline charm/packages.txt:7]: could add `gettext-base` here  to use `envsubst` :)

File: charm/serial-vault/config.yaml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    portSigning:  
+        description: "Port for the service mode=signing"
+        type: string
+        default: "8080"
+    dbdriver:
+        description: "database driver"
+        type: string
+        default: "postgres"
+    keystore:
+        description: "Signing key store, can be: databse, filesystem, memory"
```
Comment (Roadmaster) [inline charm/serial-vault/config.yaml:31]: databse -> database (it's the default so it's not a huge deal/typo)

File: ols-vms.conf [hunk 9]
```diff
@@ -2,15 +2,7 @@
 
 vm.class = lxd
 vm.architecture = amd64
-# Must be xenial because the thing is deployed on xenial and we live in the past.
-vm.release = xenial
+vm.release = focal
```
Comment (Roadmaster) [inline ols-vms.conf:5]: \o/

File: charm/serial-vault/config.yaml (hunk 8 above)
Comment (glower) [reply charm/serial-vault/config.yaml:31]: fixed

File: charm/Makefile (hunk 6 above)
Comment (glower) [reply charm/Makefile]: done

**Comment by glower** [comment]
> LGTM, worth merging and trying/iterating.
> 
> A couple of comments below but no blockers.

Fixed `sed` stuff, it was ugly. 
Tested `make charm-build` and `charm-deploy` with the new changes and it looks good.

**Pull request merged**